# String Matching in Python (pandas)

By the end of this tutorial you will know:

 " how to approximately match strings and determine how similar they are by going over various examples ? "
 


file 1 : 

In [70]:
import pandas as pd
import psycopg2


master = pd.read_csv('master_kabkota.csv')
master.head(3)

id_kab        kab_kota          provinsi
0    3372  KOTA SURAKARTA       Jawa Tengah
1    3506     KAB. KEDIRI        Jawa Timur
2    7313       KAB. WAJO  Sulawesi Selatan

File 2 : 

In [71]:
coord = pd.read_csv('polusi_kabkota.csv')
coord.head(3)

name province     district_city     kab_kota    country  \
0   Banda Aceh     Aceh   Kota Banda Aceh   Banda Aceh  Indonesia   
1       Langsa     Aceh       Kota Langsa       Langsa  Indonesia   
2  Lhokseumawe     Aceh  Kota Lhokseumawe  Lhokseumawe  Indonesia   

   coordinates_latitude  coordinates_longitude        current_ts  \
0               5.54167               95.33333  18/03/2020 05:00   
1               4.46830               97.96830  18/03/2020 16:00   
2               5.18010               97.15070  18/03/2020 16:00   

  current_condition  current_humidity  
0     Broken clouds                56  
1     Broken clouds                56  
2              Rain                66

case nya di sini kita akan mencocokan dua file, untuk mengambil semua informasi di table nya yaitu id_kab, coordinates_latitude dan coordinates_longitude 

ket : 
1. master : 514 row
2. coord : 3440 row

In [72]:
coord['kab_kota']=coord['kab_kota'].str.upper()  #mengubah format kolom kab_kota menjadi UPPERCASE
df = coord.merge(master, on='kab_kota', how='left') #merge 2 dataframe menggunakan kab_kota
df.head(3)

name province     district_city     kab_kota    country  \
0   Banda Aceh     Aceh   Kota Banda Aceh   BANDA ACEH  Indonesia   
1       Langsa     Aceh       Kota Langsa       LANGSA  Indonesia   
2  Lhokseumawe     Aceh  Kota Lhokseumawe  LHOKSEUMAWE  Indonesia   

   coordinates_latitude  coordinates_longitude        current_ts  \
0               5.54167               95.33333  18/03/2020 05:00   
1               4.46830               97.96830  18/03/2020 16:00   
2               5.18010               97.15070  18/03/2020 16:00   

  current_condition  current_humidity  id_kab provinsi  
0     Broken clouds                56     NaN      NaN  
1     Broken clouds                56     NaN      NaN  
2              Rain                66     NaN      NaN

In [73]:
df['id_kab'].isnull().sum() #jumlah kolom id_kab yang null

3440

tenyata dari 3440 data, tidak ada satupun nama kab_kota yang matching, oke dri pola nya kelihatan bahwa df1 masih terdapat kata "kab" dan "kota", maka proses selanjutnya adalah cleansing data nya (df1) 

In [74]:
master['kab_kota'] = master['kab_kota'].str.replace(r'(KAB. |KOTA )','')
master.head(3)

id_kab   kab_kota          provinsi
0    3372  SURAKARTA       Jawa Tengah
1    3506     KEDIRI        Jawa Timur
2    7313       WAJO  Sulawesi Selatan

In [98]:
coord['kab_kota']=coord['kab_kota'].str.upper()  #mengubah format kolom kab_kota menjadi UPPERCASE
df = coord.merge(master, on='kab_kota', how='left') #merge 2 dataframe menggunakan kab_kota
df = df.drop_duplicates()
df

name     province               district_city  \
0                Banda Aceh         Aceh             Kota Banda Aceh   
1                    Langsa         Aceh                 Kota Langsa   
2               Lhokseumawe         Aceh            Kota Lhokseumawe   
3                    Sabang         Aceh                 Kota Sabang   
4                  Meulaboh         Aceh        Kabupaten Aceh Barat   
5                Blangpidie         Aceh   Kabupaten Aceh Barat Daya   
6                    Jantho         Aceh        Kabupaten Aceh Besar   
7                    Calang         Aceh         Kabupaten Aceh Jaya   
8                 Tapaktuan         Aceh      Kabupaten Aceh Selatan   
9                   Singkil         Aceh      Kabupaten Aceh Singkil   
10               Karangbaru         Aceh      Kabupaten Aceh Tamiang   
11                 Takengon         Aceh       Kabupaten Aceh Tengah   
12                 Kutacane         Aceh     Kabupaten Aceh Tenggara   
13               Idi Rayeuk         Aceh        Kabupaten Aceh Timur   
14    Simpang Tiga Redelong         Aceh      Kabupaten Bener Meriah   
15                   Bireun         Aceh           Kabupaten Bireuen   
16             Blangkejeren         Aceh         Kabupaten Gayo Lues   
17              Suka Makmue         Aceh        Kabupaten Nagan Raya   
18                    Sigli         Aceh             Kabupaten Pidie   
19                 Sinabang         Aceh          Kabupaten Simeulue   
20                Mangupura         Bali            Kabupaten Badung   
21                 Amlapura         Bali        Kabupaten Karangasem   
22               Semarapura         Bali         Kabupaten Klungkung   
23                 Denpasar         Bali               Kota Denpasar   
24                   Bangli         Bali            Kabupaten Bangli   
25                Singaraja         Bali          Kabupaten Buleleng   
26                  Gianyar         Bali           Kabupaten Gianyar   
27                   Negara         Bali          Kabupaten Jembrana   
28                  Tabanan         Bali           Kabupaten Tabanan   
29                   Serang       Banten                 Kota Serang   
...                     ...          ...                         ...   
3794                Gianyar         Bali           Kabupaten Gianyar   
3795                 Negara         Bali          Kabupaten Jembrana   
3796                Tabanan         Bali           Kabupaten Tabanan   
3797                 Serang       Banten                 Kota Serang   
3798                 Serang       Banten                 Kota Serang   
3799              Tangerang       Banten              Kota Tangerang   
3800              Tangerang       Banten              Kota Tangerang   
3801        South Tangerang       Banten      Kota Tangerang Selatan   
3802          Rangkasbitung       Banten             Kabupaten Lebak   
3803             Pandeglang       Banten        Kabupaten Pandeglang   
3804                 Ciruas       Banten            Kabupaten Serang   
3805                 Ciruas       Banten            Kabupaten Serang   
3806              Tigaraksa       Banten         Kabupaten Tangerang   
3807              Tigaraksa       Banten         Kabupaten Tangerang   
3808               Bengkulu     Bengkulu               Kota Bengkulu   
3809                  Manna     Bengkulu  Kabupaten Bengkulu Selatan   
3810          Karang Tinggi     Bengkulu   Kabupaten Bengkulu Tengah   
3811             Argamakmur     Bengkulu    Kabupaten Bengkulu Utara   
3812               Bintuhan     Bengkulu              Kabupaten Kaur   
3813              Kepahiang     Bengkulu         Kabupaten Kepahiang   
3814               Mukomuko     Bengkulu          Kabupaten Mukomuko   
3815                  Curup     Bengkulu     Kabupaten Rejang Lebong   
3816                   Tais     Bengkulu            Kabupaten Seluma   
3817                Jakarta  DKI Jakarta                     Ja

In [76]:
df['id_kab'].isnull().sum() #jumlah kolom id_kab yang null

387

setelah dicleansing data nya kita ulangi lagi proses sebelumnya, dan voila terdapat 387 kab_kota yang blm termatching, baiklah coba kita lihat datanya seperti apa 

In [99]:
df_notnull = df[df['id_kab'].notnull()] #df yang id_kab nya tidak null
df1 = df[df['id_kab'].isnull()] #df yang id_kab nya null 
df1.head(3)

name     province       district_city  kab_kota    country  \
46  Mukomuko     Bengkulu  Kabupaten Mukomuko  MUKOMUKO  Indonesia   
49   Jakarta  DKI Jakarta             Jakarta   JAKARTA  Indonesia   
55    Siulak        Jambi   Kabupaten Kerinci   KERINCI  Indonesia   

    coordinates_latitude  coordinates_longitude        current_ts  \
46             -2.568900              101.11180  18/03/2020 16:00   
49             -6.236704              106.79324  18/03/2020 17:00   
55             -1.946670              101.33417  18/03/2020 16:00   

   current_condition  current_humidity  id_kab provinsi  
46              Rain                76     NaN      NaN  
49  Scattered clouds                83     NaN      NaN  
55              Rain                95     NaN      NaN

In [61]:
#import module 
from fuzzywuzzy import process

#define list 
kab_similarity=[]
similarity=[]
prov_similarity=[]
id_kab_similarity=[]

for i in df1.kab_kota:
    Ratios = process.extract(i, master.kab_kota, limit=1 )
    kab_similarity.append(Ratios[0][0])
    prov_similarity.append(master.loc[master['kab_kota'] == Ratios[0][0], 'provinsi'].iloc[0])
    similarity.append(Ratios[0][1]) 
    id_kab_similarity.append(master.loc[master['kab_kota'] == Ratios[0][0], 'id_kab'].iloc[0])

In [92]:
df1['kab_similarity']=kab_similarity
df1['id_kab_similarity'] = id_kab_similarity
df1['prov_similarity']=prov_similarity
df1['similarity']=similarity
# df1
df2 = df1[['province','kab_kota','kab_similarity','id_kab_similarity','prov_similarity','similarity']]
df2

c:\python27\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.
c:\python27\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  
c:\python27\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.

province                          kab_kota  \
46                     Bengkulu                          MUKOMUKO   
49                  DKI Jakarta                           JAKARTA   
55                        Jambi                           KERINCI   
57                        Jambi                       SUNGAIPENUH   
58                        Jambi                       BATANG HARI   
60                        Jambi                          MERANGIN   
61                        Jambi                       MUARO JAMBI   
225           Kalimantan Tengah                     PALANGKA RAYA   
249   Kepulauan Bangka Belitung                     PANGKALPINANG   
324         Nusa Tenggara Timur              TIMOR TENGAH SELATAN   
346                       Papua                PEGUNUNGAN BINTANG   
359                 Papua Barat                            FAKFAK   
370                        Riau                       ROKAN HILIR   
378                        Riau                         PELALAWAN   
396            Sulawesi Selatan                         PARE-PARE   
407            Sulawesi Selatan          PANGKAJENE DAN KEPULAUAN   
417             Sulawesi Tengah                      TOJO UNA-UNA   
418             Sulawesi Tengah                         TOLI-TOLI   
434              Sulawesi Utara         BOLAANG MONGONDOW SELATAN   
435              Sulawesi Utara           BOLAANG MONGONDOW TIMUR   
436              Sulawesi Utara           BOLAANG MONGONDOW UTARA   
437              Sulawesi Utara  KEPULAUAN SIAU TAGULANDANG BIARO   
473            Sumatera Selatan         OGAN KOMERING ULU SELATAN   
474            Sumatera Selatan        PENUKAL ABAB LEMATANG ILIR   
475            Sumatera Selatan                      LUBUKLINGGAU   
479            Sumatera Selatan           OGAN KOMERING ULU TIMUR   
480              Sumatera Utara                            ASAHAN   
481              Sumatera Utara                         BATU BARA   
482              Sumatera Utara                             DAIRI   
483              Sumatera Utara                      DELI SERDANG   
...                         ...                               ...   
3310                   Bengkulu                          MUKOMUKO   
3313                DKI Jakarta                           JAKARTA   
3319                      Jambi                           KERINCI   
3366                   Bengkulu                          MUKOMUKO   
3369                DKI Jakarta                           JAKARTA   
3375                      Jambi                           KERINCI   
3422                   Bengkulu                          MUKOMUKO   
3425                DKI Jakarta                           JAKARTA   
3431                      Jambi                           KERINCI   
3478                   Bengkulu                          MUKOMUKO   
3481                DKI Jakarta                           JAKARTA   
3487                      Jambi                           KERINCI   
3534                   Bengkulu                          MUKOMUKO   
3537                DKI Jakarta                           JAKARTA   
3543                      Jambi                           KERINCI   
3590                   Bengkulu                          MUKOMUKO   
3593                DKI Jakarta                           JAKARTA   
3599                      Jambi                           KERINCI   
3646                   Bengkulu                          MUKOMUKO   
3649                DKI Jakarta                           JAKARTA   
3655                      Jambi                           KERINCI   
3702                   Bengkulu                          MUKOMUKO   
3705                DKI Jakarta                           JAKARTA   
3711                      Jambi                           KERINCI   
3758                   Bengkulu                          MUKOMUKO   
3761                DKI Jakarta                           JAKARTA   
3767                 

terlihat hasil nya, ternyata beberpa typo yng ditemukan adalah salah penggunaan spasi. Dan telihat angka hasil similarity nya > 90%, artinya hasilnya similarity nya optimal, lanjut untuk menggabungkan dengan table sebelumnya 

In [93]:
df3 = df1[['name','province','district_city','kab_kota','country','coordinates_latitude','coordinates_longitude','current_ts','current_condition','current_humidity','id_kab_similarity','prov_similarity']]
df3

name                   province  \
46             Mukomuko                   Bengkulu   
49              Jakarta                DKI Jakarta   
55               Siulak                      Jambi   
57         Sungai Penuh                      Jambi   
58         Muara Bulian                      Jambi   
60               Bangko                      Jambi   
61              Sengeti                      Jambi   
225        Palangkaraya          Kalimantan Tengah   
249       Pangkalpinang  Kepulauan Bangka Belitung   
324                 Soe        Nusa Tenggara Timur   
346             Oksibil                      Papua   
359              Fakfak                Papua Barat   
370       Bagansiapiapi                       Riau   
378   Pangkalan Kerinci                       Riau   
396            Parepare           Sulawesi Selatan   
407          Pangkajene           Sulawesi Selatan   
417              Ampana            Sulawesi Tengah   
418            Tolitoli            Sulawesi Tengah   
434            Molibagu             Sulawesi Utara   
435             Tutuyan             Sulawesi Utara   
436              Boroko             Sulawesi Utara   
437         Ondong Siau             Sulawesi Utara   
473           Muara Dua           Sumatera Selatan   
474          Talang Ubi           Sumatera Selatan   
475        Lubuklinggau           Sumatera Selatan   
479           Martapura           Sumatera Selatan   
480             Kisaran             Sumatera Utara   
481           Limapuluh             Sumatera Utara   
482          Sidikalang             Sumatera Utara   
483         Lubuk Pakam             Sumatera Utara   
...                 ...                        ...   
3310           Mukomuko                   Bengkulu   
3313            Jakarta                DKI Jakarta   
3319             Siulak                      Jambi   
3366           Mukomuko                   Bengkulu   
3369            Jakarta                DKI Jakarta   
3375             Siulak                      Jambi   
3422           Mukomuko                   Bengkulu   
3425            Jakarta                DKI Jakarta   
3431             Siulak                      Jambi   
3478           Mukomuko                   Bengkulu   
3481            Jakarta                DKI Jakarta   
3487             Siulak                      Jambi   
3534           Mukomuko                   Bengkulu   
3537            Jakarta                DKI Jakarta   
3543             Siulak                      Jambi   
3590           Mukomuko                   Bengkulu   
3593            Jakarta                DKI Jakarta   
3599             Siulak                      Jambi   
3646           Mukomuko                   Bengkulu   
3649            Jakarta                DKI Jakarta   
3655             Siulak                      Jambi   
3702           Mukomuko                   Bengkulu   
3705            Jakarta                DKI Jakarta   
3711             Siulak                      Jambi   
3758           Mukomuko                   Bengkulu   
3761            Jakarta                DKI Jakarta   
3767             Siulak                      Jambi   
3814           Mukomuko                   Bengkulu   
3817            Jakarta                DKI Jakarta   
3823             Siulak                      Jambi   

                                   district_city  \
46                            Kabupaten Mukomuko   
49                                       Jakarta   
55                             Kabupaten Kerinci   
57                              Kota Sungaipenuh   
58                          Kabupaten Batanghari   
60                            Kabupaten Merangin   
61                         Kabupaten Muaro Jambi   
225                           Kota Palangka Raya   
249                           Kota Pangkalpinang   
324               Kabupaten Timor Tengah Selatan   
346                 Kabupaten Pegunungan Bintang   
359                             Kabupaten Fakf

In [94]:
df3 = df3.rename(columns={'id_kab_similarity':'id_kab','prov_similarity':'provinsi'})

In [95]:
result = []
result = pd.concat([df_notnull,df3])
result

name     province              district_city  \
0                Banda Aceh         Aceh            Kota Banda Aceh   
1                    Langsa         Aceh                Kota Langsa   
2               Lhokseumawe         Aceh           Kota Lhokseumawe   
3                    Sabang         Aceh                Kota Sabang   
4                  Meulaboh         Aceh       Kabupaten Aceh Barat   
5                Blangpidie         Aceh  Kabupaten Aceh Barat Daya   
6                    Jantho         Aceh       Kabupaten Aceh Besar   
7                    Calang         Aceh        Kabupaten Aceh Jaya   
8                 Tapaktuan         Aceh     Kabupaten Aceh Selatan   
9                   Singkil         Aceh     Kabupaten Aceh Singkil   
10               Karangbaru         Aceh     Kabupaten Aceh Tamiang   
11                 Takengon         Aceh      Kabupaten Aceh Tengah   
12                 Kutacane         Aceh    Kabupaten Aceh Tenggara   
13               Idi Rayeuk         Aceh       Kabupaten Aceh Timur   
14    Simpang Tiga Redelong         Aceh     Kabupaten Bener Meriah   
15                   Bireun         Aceh          Kabupaten Bireuen   
16             Blangkejeren         Aceh        Kabupaten Gayo Lues   
17              Suka Makmue         Aceh       Kabupaten Nagan Raya   
18                    Sigli         Aceh            Kabupaten Pidie   
19                 Sinabang         Aceh         Kabupaten Simeulue   
20                Mangupura         Bali           Kabupaten Badung   
21                 Amlapura         Bali       Kabupaten Karangasem   
22               Semarapura         Bali        Kabupaten Klungkung   
23                 Denpasar         Bali              Kota Denpasar   
24                   Bangli         Bali           Kabupaten Bangli   
25                Singaraja         Bali         Kabupaten Buleleng   
26                  Gianyar         Bali          Kabupaten Gianyar   
27                   Negara         Bali         Kabupaten Jembrana   
28                  Tabanan         Bali          Kabupaten Tabanan   
29                   Serang       Banten                Kota Serang   
...                     ...          ...                        ...   
3310               Mukomuko     Bengkulu         Kabupaten Mukomuko   
3313                Jakarta  DKI Jakarta                    Jakarta   
3319                 Siulak        Jambi          Kabupaten Kerinci   
3366               Mukomuko     Bengkulu         Kabupaten Mukomuko   
3369                Jakarta  DKI Jakarta                    Jakarta   
3375                 Siulak        Jambi          Kabupaten Kerinci   
3422               Mukomuko     Bengkulu         Kabupaten Mukomuko   
3425                Jakarta  DKI Jakarta                    Jakarta   
3431                 Siulak        Jambi          Kabupaten Kerinci   
3478               Mukomuko     Bengkulu         Kabupaten Mukomuko   
3481                Jakarta  DKI Jakarta                    Jakarta   
3487                 Siulak        Jambi          Kabupaten Kerinci   
3534               Mukomuko     Bengkulu         Kabupaten Mukomuko   
3537                Jakarta  DKI Jakarta                    Jakarta   
3543                 Siulak        Jambi          Kabupaten Kerinci   
3590               Mukomuko     Bengkulu         Kabupaten Mukomuko   
3593                Jakarta  DKI Jakarta                    Jakarta   
3599                 Siulak        Jambi          Kabupaten Kerinci   
3646               Mukomuko     Bengkulu         Kabupaten Mukomuko   
3649                Jakarta  DKI Jakarta                    Jakarta   
3655                 Siulak        Jambi          Kabupaten Kerinci   
3702               Mukomuko     Bengkulu         Kabupaten Mukomuko   
3705                Jakarta  DKI Jakarta                    Jakarta   
3711                 Siulak        Jambi          Kabupaten Kerinci   
3758               Mukomuko     Bengkulu       

In [100]:
result['id_kab'].isnull().sum() #jumlah kolom id_kab yang null

0

DONE !!